hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateNftsHandler.java

```diff
@@ -0,0 +1,140 @@
+import com.hedera.hapi.node.token.TokenUpdateNftsTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import com.hedera.node.app.service.token.impl.WritableNftStore;
+import com.hedera.node.app.service.token.impl.validators.TokenAttributesValidator;
+import com.hedera.node.app.spi.fees.FeeContext;
+import com.hedera.node.app.spi.fees.Fees;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.HandleException;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Provides the state transition for an NFT collection update.
+ */
+@Singleton
+public class TokenUpdateNftsHandler implements TransactionHandler {
+    private final TokenAttributesValidator validator;
+
+    @Inject
+    public TokenUpdateNftsHandler(@NonNull final TokenAttributesValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+        final var op = txn.tokenUpdateNftsOrThrow();
+        validateTruePreCheck(op.hasToken(), INVALID_TOKEN_ID);
+        validateTrue(!op.serialNumbers().isEmpty(), MISSING_SERIAL_NUMBERS);
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        pureChecks(context.body());
+    }
+
+    @Override
+    public void handle(@NonNull HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txnBody = context.body();
+        final var op = txnBody.tokenUpdateNftsOrThrow();
+        final var token = op.tokenOrThrow();
+
+        validateSemantics(context, op);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+
+        // Wrap in Set to de-duplicate serial numbers
+        final var nftSerialNums = new LinkedHashSet<>(op.serialNumbers());
+        validateTrue(nftSerialNums.size() <= nftStore.sizeOfState(), INVALID_NFT_ID);
+        updateNftMetadata(nftSerialNums, nftStore, token, op);
+    }
+
+    private void updateNftMetadata(
+            @NonNull final Set<Long> nftSerialNums,
+            @NonNull final WritableNftStore nftStore,
+            @NonNull final TokenID tokenNftId,
+            @NonNull final TokenUpdateNftsTransactionBody op) {
+        final var hasMetadata = op.hasMetadata();
+        final var metadata = hasMetadata ? op.metadata() : Bytes.EMPTY;
```

Comment:
Is it possible to clear a `nft.metadata` or why would we not fail if metadata is empty? Is it because in the future we might allow updating of some other field so we want to support the no-op transaction of updating without a metadata?

Reply:
As discussed offline, we should not allow clear since nft.metadata is required. We should throw if empty.

Reply:
Makes sense. Good catch!

Reply:
Fixed. Thanks @steven-sheehy 👍

Reply:
I assume we did not enforce it in Mono either, right? Let's create an issue and fix it after modularization. Don't want to break diff testing.

Reply:
We did not enforce in mono also. 
Makes sense. Will do